pkg/typesystem/typesystem.go

```diff
@@ -379,6 +411,52 @@ func (t *TypeSystem) UsersetCanFastPath(relationReferences []*openfgav1.Relation
 	return true
 }
 
+// recursiveUsersetNodeCanFastpath is a helper function to determine whether the node (object#relation) has
+// - one edge back to itself
+// - other edges lead directly to terminal types (not union/intersection/exclusion).
+func (t *TypeSystem) recursiveUsersetNodeCanFastpath(curAuthorizationModelNode *graph.AuthorizationModelNode, userType string) bool {
+	hasNodeBackToItself := false
+	hasCorrectTerminalType := false
+
+	// TODO: Optimize by memorizing the list of childrenNodes
+	childrenNodes := t.authorizationModelGraph.To(curAuthorizationModelNode.ID())
```

Comment:
Nitpick: I prefer the naming `neighborNodes`. `childrenNodes` makes it sound like the graph is a tree, and it is not always a tree.